ubuntu_image/tests/test_builder.py

```diff
@@ -826,6 +826,121 @@ def test_make_disk_with_parts(self):
+                name='beta',
+                type='bb',
+                size=MiB(1),
+                offset=MiB(4),
+                offset_write=200,
+                )
+            part2 = SimpleNamespace(
+                name='gamma',
+                type='mbr',
+                size=MiB(1),
+                offset=0,
+                offset_write=None,
+                )
+            volume = SimpleNamespace(
+                # gadget.yaml appearance order which does not match the disk
+                # offset order.  LP: #1642999
+                structures=[part1, part0, part2],
+                schema=VolumeSchema.gpt,
+                )
+            state.gadget = SimpleNamespace(
+                volumes=dict(volume1=volume),
+                )
+            # Set up images for the targeted test.  These represent the image
+            # files that would have already been crafted to write to the disk
+            # in early (untested here) stages of the state machine.
+            part0_img = os.path.join(state.images, 'part0.img')
+            with open(part0_img, 'wb') as fp:
+                fp.write(b'\1' * 11)
+            part1_img = os.path.join(state.images, 'part1.img')
+            with open(part1_img, 'wb') as fp:
+                fp.write(b'\2' * 12)
+            part2_img = os.path.join(state.images, 'part2.img')
+            with open(part2_img, 'wb') as fp:
+                fp.write(b'\3' * 13)
+            state.root_img = os.path.join(state.images, 'root.img')
+            state.rootfs_size = MiB(1)
+            with open(state.root_img, 'wb') as fp:
+                fp.write(b'\4' * 14)
+            state.boot_images = [part1_img, part0_img, part2_img]
+            # Create the disk.
+            next(state)
+            # Verify some parts of the disk.img's content.  First, that we've
+            # written the part offsets at the right place.y
+            with open(state.disk_img, 'rb') as fp:
+                # Part 0's offset is written at position 100.
+                fp.seek(100)
+                # Read a 32-bit little-ending integer.  Remember
+                # struct.unpack() always returns tuples, and the values are
+                # written in sector units, which are hard-coded as 512 bytes.
+                offset = unpack('<I', fp.read(4))[0]
+                self.assertEqual(offset, MiB(2) / 512)
+                # Part 1's offset is written at position 200.
+                fp.seek(200)
+                offset = unpack('<I', fp.read(4))[0]
+                self.assertEqual(offset, MiB(4) / 512)
+                # Part 0's image lives at MiB(2).
+                fp.seek(MiB(2))
+                self.assertEqual(fp.read(15), b'\1' * 11 + b'\0' * 4)
+                # Part 1's image lives at MiB(4).
+                fp.seek(MiB(4))
+                self.assertEqual(fp.read(15), b'\2' * 12 + b'\0' * 3)
+                # Part 2's image is an MBR so it must live at offset 0.
```

Comment:
Yeah, good idea.  We have lots of tests for implicit offsets, but not one specifically for testing the written partitions.  I'll add that in this branch.